File: macro_pipeline/core/pipeline.py
import io
import zipfile
from pathlib import Path
from typing import Protocol, Sequence

import polars as pl


class DataLoader(Protocol):
    def load(self, file_path: Path) -> pl.DataFrame: ...


class DataTransformer(Protocol):
    def transform(self, df: pl.DataFrame) -> pl.DataFrame: ...


class DataSaver(Protocol):
    def save(self, df: pl.DataFrame, file_path: Path) -> None: ...


class CsvLoader(DataLoader):
    def load(self, file_path: Path) -> pl.DataFrame:
        df = pl.read_csv(file_path)
        df = df.rename({df.columns[0]: "year"})
        return df


class InsertIndicatorAndSwapCols(DataTransformer):
    """
    Insert last column name as 'indicator', keep its values as 'value',
    while preserving 'year'.
    Final order: ['year', 'indicator', 'value'].
    """

    def transform(self, df: pl.DataFrame) -> pl.DataFrame:
        if df.width < 2:
            return df

        year_col = df.columns[0]
        value_col = df.columns[-1]

        return (
            df.select([year_col, value_col])
            .rename({year_col: "year", value_col: "value"})
            .with_columns(
                pl.col("value").cast(pl.Float64),
                pl.lit(value_col).alias("indicator"),
            )
            .select(["year", "indicator", "value"])
        )


class InMemoryZipSaver(DataSaver):
    """
    Saves a Polars DataFrame directly into a ZIP file in memory
    without creating intermediate CSV files on disk.
    """

    def save(self, df: pl.DataFrame, file_path: Path) -> None:
        csv_buffer = io.StringIO()
        df.write_csv(csv_buffer, include_header=True)
        csv_buffer.seek(0)

        with zipfile.ZipFile(
            file_path.with_suffix(".zip"),
            mode="w",
            compression=zipfile.ZIP_DEFLATED,
        ) as archive:
            archive.writestr(file_path.name, csv_buffer.getvalue())


class DataPipeline:
    def __init__(
        self,
        loaders: Sequence[DataLoader],
        transformers: Sequence[DataTransformer],
        saver: DataSaver,
    ) -> None:
        self.loaders = loaders
        self.transformers = transformers
        self.saver = saver

    def run(self, input_paths: Sequence[Path], output_path: Path) -> None:
        frames: list[pl.DataFrame] = []

        for path in input_paths:
            for loader in self.loaders:
                chunk = loader.load(path)
                for transformer in self.transformers:
                    chunk = transformer.transform(chunk)
                frames.append(chunk)

        df = pl.concat(frames, how="vertical", rechunk=True)

        self.saver.save(df, output_path)

File: src/core/pipeline.rs
use std::fs::File;
use std::io::Write;
use std::path::Path;
 
use anyhow::Result;
use polars::prelude::*;

use ::zip::write::FileOptions;
use ::zip::ZipWriter;
 

pub trait DataLoader {
    fn load(&self, path: &Path) -> Result<LazyFrame>;
}
 
pub trait DataTransformer {
    fn transform(&self, lf: LazyFrame) -> LazyFrame;
}
 
pub trait DataSaver {
    fn save(&self, df: &mut DataFrame, path: &Path) -> Result<()>;
}
 

pub struct CsvLoader;
 
impl DataLoader for CsvLoader {
    fn load(&self, path: &Path) -> Result<LazyFrame> {
        Ok(
            LazyCsvReader::new(path)
                .with_has_header(true)
                .finish()?
        )
    }
}
 

pub struct InsertIndicatorAndSwapCols;
 
impl DataTransformer for InsertIndicatorAndSwapCols {
    fn transform(&self, lf: LazyFrame) -> LazyFrame {
        // We need column names >> must inspect schema
        let schema = lf.schema().expect("schema available");
        let cols = schema.iter_names().collect::<Vec<_>>();
 
        if cols.len() < 2 {
            return lf;
        }
 
        let year_col = cols[0].to_string();
        let value_col = cols.last().unwrap().to_string();
 
        lf.select([
            col(&year_col).alias("year"),
            col(&value_col)
                .cast(DataType::Float64)
                .alias("value"),
            lit(value_col.clone()).alias("indicator"),
        ])
    }
}
 

pub struct ZipCsvSaver;
 
impl DataSaver for ZipCsvSaver {
    fn save(&self, df: &mut DataFrame, path: &Path) -> Result<()> {
        let zip_path = path.with_extension("zip");
 
        let file = File::create(zip_path)?;
        let mut zip = ZipWriter::new(file);
 
        let options = FileOptions::default();
 
        zip.start_file(
            path.file_name().unwrap().to_string_lossy(),
            options,
        )?;
 
        let mut buffer: Vec<u8> = Vec::new();

        CsvWriter::new(&mut buffer)
            .include_header(true)
            .finish(df)?;
 
        zip.write_all(&buffer)?;
        zip.finish()?;
 
        Ok(())
    }
}
 

pub struct DataPipeline<L, T, S>
where
    L: DataLoader,
    T: DataTransformer,
    S: DataSaver,
{
    pub loaders: Vec<L>,
    pub transformers: Vec<T>,
    pub saver: S,
}
 
impl<L, T, S> DataPipeline<L, T, S>
where
    L: DataLoader,
    T: DataTransformer,
    S: DataSaver,
{
    pub fn run(&self, input_paths: &[&Path], output_path: &Path) -> Result<()> {
        let mut plans: Vec<LazyFrame> = Vec::new();
 
        for path in input_paths {
            for loader in &self.loaders {
                let mut lf = loader.load(path)?;
 
                for transformer in &self.transformers {
                    lf = transformer.transform(lf);
                }
 
                plans.push(lf);
            }
        }
 
        // Combine all lazy frames
        let combined = concat(plans, UnionArgs::default())?;
 
        // Execute once
        let mut df = combined.collect()?;
 
        self.saver.save(&mut df, output_path)?;
 
        Ok(())
    }
}
